### DISCUSSION TITLE
What would an implementation of this framework look like?

### DISCUSSION THREAD

**Discussion by mdubinko** [daveshap/ACE_Framework#2] [General]
I'm interested an anyone's thoughts on this. What would an implementation look like? Is this purely conceptual a la OSI layers, or could all or parts of it get implemented as software+template components? (And in that case, what would a reference implementation look like?)

**Comment by daveshap** [comment]
We're working on several demo implementations.

**Reply by Qriscuit** [reply]
Do you have a clear idea of what the architecture of this software looks like ? if so a UML diagram will be really helpful. And when do you think will we have the very first implementation by ?

**Reply by daveshap** [reply]
This is a framework like OSI. You should read the framework or the paper first.

**Comment by daveshap** [comment]
> I'm interested an anyone's thoughts on this. What would an implementation look like? Is this purely conceptual a la OSI layers, or could all or parts of it get implemented as software+template components? (And in that case, what would a reference implementation look like?)

Looks like the team wants to build HER first (from the movie)

**Reply by jjtolton** [reply]
I'm curious specifically about the flow control. For instance, do the layers act in parallel? Or sequentially with priority? For instance, the case of the morality layer issuing a sigkill is interesting. I'm also interested if lower layers can have access to higher layer scratch pad space -- ie, if the task layer (for example) determines the best way to save battery power is by killing off all the pesky cycles going into morality adjudication, it could (in a poorly configured agent) rewrite the morality framework.  Of course, that shouldn't be possible given the way directives flow, but the point of the question is "can every low layer operate independently of the higher layers simply by removing them"?  if so, this has an interesting feeling of LLM-style "macro expansion" from higher layers to lower layers.

**Reply by daveshap** [reply]
Parallel eventually like any SOA or microservices architecture

**Reply by jjtolton** [reply]
That makes sense. From an implementation perspective, I'm curious what your thoughts are on the bus.  Would it be an append-only log, or a ring buffer?  Or would you need a writer to mediate?

![image](https://github.com/daveshap/ACE_Framework/assets/10394611/62b95fa7-944a-4c02-8c77-be6491d63f63)

The bus makes sense, but I imagine these things trampling on each other if not managed carefully. 

I can also see a priority queue being used to hold proposed actions, so that higher layer actions are taken first.  And perhaps software transactional memory or command pattern / unit-of-work / etc being used to ensure that edits to the bus are valid (i.e., not updating stale data). 

Also I'm curious if you're imagining this (from a CAP perspective) as strongly available / not partition tolerant?

**Reply by daveshap** [reply]
Look up AMQP, this is a solved problem

**Reply by dedumas** [reply]
Edit: didn't see AMQP reference as this response was being authored

One option for a simple bus implementation might be to have the layers be in a linked list with one another, passing trains of thought (telemetry and control could use the same linkage in opposite directions). I imagine two possibilities for when a layer might send a thought, or train of thoughts, along to the next "bus stop" (next layer in the LL).

(1) This layer has an original thought, maybe its steady state has a timer fire, or input from a sensor, but the critical distinction here is that it is **not** triggered by any pre-existing train of thought coming _in_ on the bus. In this case, the layer creates its own queue which represents a "train" (of size 1) of thought, it's writing its thought for telemetry for example to pass "up" to the higher layers

(2) is predicated on _(1)_ having occurred at some prior layer and each layer has the capability of being provided a train of thought. The first action is to review the thoughts already in this queue train of thought, second action is to decide whether anything belongs on the _other_ bus (i.e. if this is the control bus it may result in an original thought initiating at this layer _(1)_ for telemetry bus or vice versa). Third action is to decide whether anything needs to be added to _this_ train of thought. If yes, add this layers thought, if no, don't, but no matter what final action is to call the next layer in the linked list, passing train of thought so that _(2)_ can occur there. Finally, when the linked list of layers ends that train of thought is discarded and considered completely handled. 

This implementation offers a way for each layer to contextualize the thoughts in a train and delineate between discrete intentions. I think it would also be parallelizable relatively easily.

In serial, having prioritized trains of thought, or thoughts, feels error prone and dangerous. Assuming layers can process thoughts in parallel it might be more easily handled. But my thinking is that "priority" would actually be handled at the lower layer implementations, more analogous to a brain with the prioritization happening at the Executive Function layer. As I say that I'm reminded of considering whether this framework should have room for a "subconscious" somewhere. Assuming this agent were embodied at any point there is probably value to having a layer or system which can execute some functions without "thinking" about it. i.e. you have a humanoid robot falling and it will probably be too slow to have that chain of thought go all the way up and come back down as a "control" signal to catch itself. Sure, some of this can / will be handled at the hardware layer itself, but while I don't have a good example I'm imagining these _types_ of "thoughtless" responses will probably be necessary.

**Reply by jjtolton** [reply]
It seems to me that AMQP solves part of the problem, but it really doesn't solve the "layer priority consensus" aspect, and it doesn't solve trampling edits to the bus in a stale data concurrent write scenario.  I'm sure you've thought of this already and I'm sure it's also a solved problem, it's just not clear to me the method being used.

**Reply by daveshap** [reply]
> It seems to me that AMQP solves part of the problem, but it really doesn't solve the "layer priority consensus" aspect, and it doesn't solve trampling edits to the bus in a stale data concurrent write scenario. I'm sure you've thought of this already and I'm sure it's also a solved problem, it's just not clear to me the method being used.

Oh I gotcha. So the way AMQP works is that you have producers and consumers. You can also create channels and clusters so whenever you produce a message, you know exactly who it's going to. You don't "edit" messages on the bus. So for instance, on the north bus, layer 3 may publish (produce) some messages that are directly consumed by Layer 2 and Layer 1.

@jjtolton

**Reply by jjtolton** [reply]
Ok, that makes more sense. What I meant by "edit" is, each layer has at least a "materialized" state -- the simplest version of this state would be plain text, since that is consumable by the LLMs. Additionally, each layer has some notion of the materialized state of the other layers (knowledge of lower layers comes from the northbound bus, knowledge of higher layers comes from the southbound bus).  So, in the concrete example I'm thinking of, the top layer made get a message from the northbound bus that bottom layer is doing something naughty.  Therefore, top layer sends a [STOP ACTIVITY] command.  However, it's possible that bottom layer has moved on to a different activity already, and the command terminates the NEW bottom layer activity.  

So overall I'm just asking what the plan is to tackle stale data/state and race conditions.  I think what I'm seeing is an "eventually consistent" distributed model, and I'm wondering if there's a plan for coordination and ordering to prevent that sort of thing or is this a "TBD" situation?

**Reply by jjtolton** [reply]
I'm also just making some wild assumptions that the messages include either commands or state deltas, and typically not the whole state (of the producer).  But maybe it does send the whole state?  Lot's of really interesting possibilities here and very curious to see where you and your team go with it.  Y'all are on the tip of the spear, for sure.

**Reply by samgriek** [reply]
IMO the layers are quite light weight and the heavy lifting happens in the LLM itself (what do you think?). I implemented them all as docker images and went with NATS because of protobuf + Golang + large message support.  It's just a hello world.  Basically the `aspirant` sends a `hello` to the `strategist` and each subsequent layer sends the `hello` down.  The prosecutor reverts with `hello to you` which goes right back up to the top.  I left out the aspirant logs because it subscribes to everything.  I didn't set up the `auditor` yet, but it will definitely be needed because the `aspirant` should be busy figuring if the actions are well aligned with the mission (presumably).  Certainly it can't hurt to have more than one entity looking out for human interests in a powerful tool like this.  Logs:

```
aspirant-1    | INFO:base.settings.send_hello=True
aspirant-1    | INFO:base 'hello' to southbound.strategist
strategist-1  | INFO:__main__:received hello from southbound.strategist
strategist-1  | INFO:__main__:published 'hello' to southbound.agent
agent-1       | INFO:__main__:received hello from southbound.agent
agent-1       | INFO:__main__:published 'hello' to southbound.executive
executive-1   | INFO:__main__:received hello from southbound.executive
executive-1   | INFO:__main__:published 'hello' to southbound.controller
controller-1  | INFO:__main__:received hello from southbound.controller
controller-1  | INFO:__main__:published 'hello' to southbound.prosecutor
prosecutor-1  | INFO:__main__:received hello from southbound.prosecutor
prosecutor-1  | INFO:__main__:revert northbound 'hello' to northbound.controller
controller-1  | INFO:__main__:northbound message => hello to you
executive-1   | INFO:__main__:northbound message => hello to you
agent-1       | INFO:__main__:northbound message => hello to you
strategist-1  | INFO:__main__:northbound message => hello to you
```

@daveshap - What do you think about REDIS for storing context and mission related data at each level (separately) so as to augment whatever LLM is being used?

Here is the log below... maybe this helps you @jjtolton ? If you have any questions about message busses, I can help clarify that for you.  I have experience with JBoss JMS, RabbitMQ, AMQP, Mosquito, Kafka, and as of today NATS.  I was thinking that TLS alone is likely not good enough for this setup and each message should be encrypted where only the intended recipient has the private key.  This way even if somehow the `prosecutor` which is the only entity that can 'write and execute' code (presumably) started to subscribe to messages intended for the other layers, it wouldn't be able to read them.

**Reply by DataBassGit** [reply]
@jjtolton It can be either depending on your usecase. Remember, this is just the framework. In most cases, full state logging is probably the best bet, so long as you are categorizing datapoints properly. However you choose to store your data, you can return on the data needed for whichever layer agent it goes to. This is also good for human verification of data.

However, there needs to be infrastructure to support such a system. Since we are potentially talking about hundreds of messages per second, monitoring and maintaining such a system would be heavy on the host. For structured data queries like you would want for verification, document stores are the most light weight. (Think MongoDB) But, querying these kinds of datastores against unindexed values is also very heavy. So a query like:

```
db.collection.find({
  "metadata.tags": "layer=1", 
  "timestamp": {
    $gte: new Date("September 1, 2023 00:00:00"),
    $lt: new Date("September 1, 2023 00:10:00")
  }
})
```

would be relatively lightweight, as we would index against the layer tag and timestamps are just math operations. However, 

```
db.logs.find({
 "metadata.tags": "layer=1",
  message: { 
    $regex: /murder/
  }
})
```

is going to be a much more intense task because this is a string operation. But, that can be alleviated by offloading some of that work to the GPU. This is where vector databases shine. Because searches like this (Semantic Similarity) are done mathematically, they can actually be performed by the graphics card.

Now, you don't want to be using precious GPU on every single call, and for the most part, the AI doesn't need to always access this data, it doesn't really matter what system your bus runs on. You just devise your bus to always dump new messages into your logging system. The bus would only need to keep the very most recent messages from each layer. Therefore, your bus has plenty of room for full state, and you can let the database handle separating deltas when you scale up to thousands of inferences per second.

**Reply by jjtolton** [reply]
Yeah, very interesting.  Great discussion.  I think part of my confusion initially was around the terminology "bus".  In CE/EE terminology, bus has the connotation of something where the signal is the same at all points, and so I was initially thinking some sort of shared buffer between the layers. The message broker discussion was very clarifying, but depending on how this is implemented, now we have a consistency issue. That being said, this isn't necessarily a problem.  In nature we have several examples of an organism that experiences high (or infinite) internal latency between processing centers.  The long-neck dinosaurs were examples of the multi-brain (unless that story changed since I was in kindergarten).  Perhaps a more salient example is the patient who has undergone a [corpus callosotomy](https://en.wikipedia.org/wiki/Corpus_callosotomy) and is still able to function mostly normally despite both halves of the brain being unable to communicate at all (except indirectly -- message passing, via sight, sound, touch, etc).  

Taking a step back, it's probably better to dial in on some assumptions before charging ahead with the implementation discussion.  For instance, @DataBassGit, you've mentioned external state -- MongoDB, in this case, or a Vector database.  @samgriek you mentioned  Redis.  Zookeeper/etcd also come to mind, here. 

Some other assumptions are probably worth digging out. Are the LLMs in charge of these layers at GPT-4 level?  Are they fine-tuned?  Or could these also be GPT-3.5 or GPT-3 equivalent, with prompt-tuning only?  

As I see it, one of the major open problems with any agent system right now is keeping it on track, due to the incredible short-term memory. One thing that concerns me about sending, for instance, full state to a model that **hasn't** been fine tuned is that the model may forget who it is and assume the identity of a higher (or lower) layer by accident. Whereas, if all models were GPT-4, this would be less likely to happen.  

I think my **hope** for this framework is that a collection of well selected open source LLMs runnable on commodity hardware driving the layers, with minimal instruct fine-tuning, and some degree of retrieval augmented generation (to stay on track), could out-perform more expensive/exclusive models.

**Reply by DataBassGit** [reply]
@jjtolton No, you're getting it. The bus can literally be any system that holds information. In my example, I may not have been clear, the database stores logs of the bus state. It can also act as the bus, however. Hell, the bus can be any file on the computer, or just a variable in the app.

Memory is actually not really a problem. The browser chatbots out there right now are essentially just filling the context window with the chat history. AutoAgents often employ a database, but the implementation isn't always the best. That being said, any system with a database has infinite memory, it's just a matter of remembering the right things at the right time.

When we're talking about full state vs delta, we're talking about what's stored on the bus, not what goes to the model. Each layer agent can pull the relevant data needed from the bus. That data presumably gets injected into a prompt template. This reduces token usage, saves context window space, and facilitates better attention from the model.

**Reply by jjtolton** [reply]
@DataBassGit right on.  And this is where we get into the really nitty-gritty.  

> That being said, any system with a database has infinite memory, it's just a matter of remembering the right things at the right time.
> 
> When we're talking about full state vs delta, we're talking about what's stored on the bus, not what goes to the model. Each layer agent can pull the relevant data needed from the bus. That data presumably gets injected into a prompt template. This reduces token usage, saves context window space, and facilitates better attention from the model.

I think one of the major challenges here is getting the agent, or sub-agent-layer-boss, to be aware it can do that -- and to do that reliably.  I've been playing with Open Interpreter -- they have a very interesting take on the problem, but you can tell from the code how much they are struggling to extract possible code from non-GPT-4 models.  So turning these things into tool users -- well, let me ask -- do you think that's a significant part of this puzzle?

I actually think it's so important that it nearly begs the question.  If we can get agents to reliably use (and discover, build, reuse, etc) tools, that may obviate at least a part of the need for this framework (although the morality part of it is nice).

**Reply by DataBassGit** [reply]
@jjtolton Check out my other project, AgentForge. It's pinned on my profile. We have a pretty nifty tool use solution, but it's not been put through the ringer yet. Essentially, we select tools by semantic search. The agent deploys tools in a very similar way to how GPT-4's function usage works. And, most importantly, it's set up to work for any (instruct trained) model and can use any python library.

Like I said, hasn't been put through the ringer yet, but preliminary tests work.
